[LLP] Optimized InMemoryState lifecycle management [DPP-1137]

tudor-da opened the pull request:
**Motivation for this PR**
Currently, any Indexer restart leads to a call to `InMemoryState.initializeTo`. This triggers a full clean-up of the in-memory state:
* Flushing all the caches leading to degraded performance until re- warm-up.
* Cancelling all the Ledger API client streams
* (In the future, as envisioned) re-population of the package cache used for interface subscriptions, an operation that can take tens of seconds and even more.

As the Indexer restart can happen quite frequently in setups with choppy DB connections, the full restart of the in-memory state on each transient error can lead to degraded UX.

**Resolution**
 This PR makes actual resetting of the in-memory state conditional and exceptional. The in-memory state is restarted only if a failure could have rendered its state:
* **stale**: when the Indexer Akka streams pipeline fails just after ingesting the persisted ledger end but before the updates are forwarded to the in-memory state.
* **dirty**: when the methods mutating its objects (dispatcher, caches, stringInterningView) fails. 
* **ahead-of-the-indexer**: when an async commit fails in the Indexer after the not-yet-committed-to-persistence updates have been ingested in the in-memory fan-out.

We perform the checks explicitly:
* **stale** and **ahead-of-the-indexer** we check by comparing the ledger end cache against the re-initialization ledger end. Additionally, we check the re-initialization string interning id against the last string interned id in the view (we need to also check this as **dirty** does not cover it since the StringInterningView is updated in the `Indexer`.
* **dirty** we define the `_dirty` flag explicitly that is set on failures in `InitializeTo` and `update`.

P.S. For better encapsulation of the mutable operations with the mutating objects, the objects mutating method is extracted from `InMemoryStateUpdater.flow` to `InMemoryState.update`

P.P.S. This PR does not take care of synchronizing mutating operations on the in-memory state. This synchronization is guaranteed from the outside. However, added a TODO in `InMemoryState` for enabling this synchronization internally. 

changelog_begin
changelog_end

<!--
# Pull Request Checklist

- Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- Include appropriate tests
- Set a descriptive title and thorough description
- Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- Normal production system change, include purpose of change in description
- If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.
-->

simonmaxen-da commented on ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:
```diff
@@ -42,7 +41,12 @@ final class InMemoryStateUpdater(
     ) => TransactionLogUpdate.TransactionRejected,
-    updateLedgerEnd: (Offset, Long) => Unit,
+    convertToContractStateEvents: TransactionLogUpdate => Vector[ContractStateEvent],
+    updateInMemoryState: (
```
This is nicer.

simonmaxen-da reviewed (commented):
Added some comments, but I'm not sure I fully understand this change.

simonmaxen-da commented:
I think it would be worth splitting this PR into two:
- One that addresses the places where we may serve incorrect data following failure
- One that addresses a temporary performance degradation following failure

tudor-da commented:
> I think it would be worth splitting this PR into two:
> 
> * One that addresses the places where we may serve incorrect data following failure
> * One that addresses a temporary performance degradation following failure

Extracted resetting the DispatcherState in two stages in separate PR: https://github.com/digital-asset/daml/pull/14573

tudor-da commented:
Thanks for the review @simonmaxen-da . Replacing this PR with a new one (https://github.com/digital-asset/daml/pull/14612) since this one's original comments and foundation is different. Addressed your comments in the new.
